### ISSUE TITLE
Plotting Refactor – Exploration & Requirements Gathering

### ISSUE THREAD

**Issue by VeraChristina** [ecmwf/anemoi-core#769]
### Goal
Improve maintainability, clarity, and usability of plotting callbacks across the codebase.
Scope of this Issue
This issue is for exploration and documentation only.
Implementation will be split into separate follow-up tickets once we understand the scope.
 
### Areas to Investigate
1. Code Structure & Fragile Assumptions
•	Identify tight coupling and brittle logic within existing plotting components
•	Note areas where assumptions break with changes in training, or with new plots/use cases
•	Can we support plotting from logged outputs instead of recomputing in the callbacks
2. Sharding / Distributed Training Compatibility
•	Understand how current plots behave under sharded or distributed operation
•	Identify pain points
3. Lightning Callback Integration
•	Review how callbacks interact with PyTorch Lightning internals
4. Multi-Dataset Plotting
•	Evaluate how multiple datasets are handled
•	Define requirements for effective handling of multiple datasets trainings
 
### Plot Functionality & Tools
5. Replace / Integrate Map-Based Plotting
•	Investigate using earthkit-plots for map-based outputs
•	Explore integration opportunities with Datashader where applicable
6. Plot Audit
•	Identify existing plots that may be obsolete or rarely used, e.g. long rollout plots?
•	List potential new plots needed, based on user experience or research workflows
7. Plot Performance Bottlenecks
•	Review slow plotting components, e.g. PlotSpectrum
8. CLI / Workflow Separation
•	Investigate running plots outside training:
•	CLI subcommand for validation visualisation / metric reporting
9. Asynchronous Computation
•	Revisit current setup for async plotting
 
### Outcomes Expected From This Issue
•	Documented overview of current plotting challenges
•	Concrete list of follow-up tasks + tickets
•	Prioritization suggestions (cleanup, replacement, retirement)

**Comment by anaprietonem** [comment]
Update on steps and scope of changes. 
Standardise callback configuration and runtime architecture to:

- Improve maintainability
- Enable multi-dataset flexibility
- Support future extensibility (e.g., Earthkit backend)
- Preserve current behavior during migration
- Proposed Configuration Model

**1. Runtime Settings** 

Introduce a dedicated settings section for runtime-only behavior:
```
settings:
  schedule:
    batch: 750
    epoch: 5

  backend: datashader
  projection_kind: equirectangular
  asynchronous: true

  sample_idx: 0
  per_sample: 6

  precip_and_related_fields: [tp, cp]

  colormaps:
    default:
      _target_: anemoi.training.utils.custom_colormaps.MatplotlibColormap
      name: viridis

    error:
      _target_: anemoi.training.utils.custom_colormaps.MatplotlibColormap
      name: bwr

    precip:
      _target_: anemoi.training.utils.custom_colormaps.MatplotlibColormapClevels
      clevels: ["#ffffff", "#04e9e7", "#019ff4", "#0300f4", "#02fd02", "#01c501", "#008e00", "#fdf802", "#e5bc00", "#fd9500", "#fd0000", "#d40000", "#bc0000", "#f800fd"]
      variables: ${diagnostics.plot.settings.precip_and_related_fields}

  focus_areas:
    europe:
      latlon_bbox: [30.0, -20.0, 60.0, 40.0]
    china:
      latlon_bbox: [18.0, 73.0, 54.0, 135.0]
```
This section is handled by a new PlotRuntimeSettings class.

Responsibilities:

Parse and validate runtime keys
Provide defaults and normalized values
Stay independent from callback semantics (e.g., parameters, output_steps)

**2. Callback Configuration**
Callbacks remain list-based but adopt a unified dataset interface:
```
callbacks:
  - _target_: anemoi.training.diagnostics.callbacks.plot.GraphTrainableFeaturesPlot
    dataset_settings:
      era5: {}
      cerra: {}
    every_n_epochs: ${diagnostics.plot.settings.schedule.epoch}

  - _target_: anemoi.training.diagnostics.callbacks.plot.PlotLoss
    dataset_settings:
      era5:
        parameter_groups:
          moisture: [tp, cp, tcw]
          sfc_wind: [10u, 10v]
      cerra:
        parameter_groups:
          moisture: [tp, cp]
    every_n_batches: ${diagnostics.plot.settings.schedule.batch}

  - _target_: anemoi.training.diagnostics.callbacks.plot.PlotSample
    sample_idx: ${diagnostics.plot.settings.sample_idx}
    per_sample: ${diagnostics.plot.settings.per_sample}
    every_n_batches: ${diagnostics.plot.settings.schedule.batch}
    accumulation_levels_plot: [0, 0.05, 0.1, 0.25, 0.5, 1, 1.5, 2, 3, 4, 5, 6, 7, 100]
    precip_and_related_fields: ${diagnostics.plot.settings.precip_and_related_fields}
    colormaps: ${diagnostics.plot.settings.colormaps}
    dataset_settings:
      era5:
        output_steps: ${training.multistep_output}
        parameters: [z_500, t_850, u_850, v_850, 2t, 10u, 10v, sp, tp, cp]
      cerra:
        output_steps: ${training.multistep_output}
        parameters: [z_500, t_850, u_850, v_850, 2t, tp]

  - _target_: anemoi.training.diagnostics.callbacks.plot.PlotHistogram
    sample_idx: ${diagnostics.plot.settings.sample_idx}
    every_n_batches: ${diagnostics.plot.settings.schedule.batch}
    precip_and_related_fields: ${diagnostics.plot.settings.precip_and_related_fields}
    dataset_settings:
      era5:
        output_steps: ${training.multistep_output}
        parameters: [z_500, tp, 2t, 10u, 10v]
      cerra:
        output_steps: ${training.multistep_output}
        parameters: [z_500, tp, 2t]
```
**3. Introduce CallbackContext**

Avoid passing the full config into get_callbacks().

Instead: pass a lightweight CallbackContext containing:
- runtime_settings: PlotRuntimeSettings
- Logger / artifact handlers
- Rank / world info
- Model/task plotting adapter
- Relevant training/system config

@anaprietonem  looking at it (with @HCookie )

**4. Split plot.py into Modules**
Refactor the current monolith into:
callbacks/plot/
-   base.py
-   loss.py
-   sample.py
-   histogram.py
-   spectrum.py
-   graph_features.py
-   schedule.py
-  context.py
Maintain backward compatibility via __init__.py re-exports

**5. Geospatial Submodule**
Move geospatial logic into a dedicated namespace:

diagnostics/geospatial/ or diagnostics/plot/geospatial/

Responsibilities:
- Focus area / masking logic
- Projection helpers
- Lat/lon transformations

**6. Async Execution → AsyncPlotRunner**
Extract async logic from callbacks.
Responsibilities:
- Event loop / executor lifecycle
- Error propagation
- Graceful shutdown

**7. Rendering Backend Abstraction**
Investigate if a  renderer interface can be introduced to support.
- MatplotlibRenderer
- DatashaderRenderer
-EarthkitPlotRenderer

Earthkit-plot could allow to simplify some of the plots as it already includes relevant feature such as 'focus area' cropping as highlighted by @HCookie . Harrison will start looking at how feasible is to reproduce the plots we have with earthkit-plots to understand this better

**8. Validation CLI**
Introduce:
```
anemoi-training validate
```
Idea to be able to run plotting callbacks independently with fine grained controlled of the checkpoint used.

**9. Improve Data Flow & Post-Processing** 
Current Problems

- PlotLoss recomputes loss inside callback
- Other callbacks reprocess batch data

Proposed Solutions

- Option A: Trainer Output Contract. Trainer emits plot-ready tensors/metrics
- Option B: Plot Adapter. Adapter normalizes trainer outputs once
Callbacks consume standardized payload

@VeraChristina  - will start to look at it.

10. Plot Audit -  
Take this opportunity to simplify and deprecate plots that are not being used or are no longer useful. 
- LongRolloutPlots - proposal to deprecate ( UPDATE - now deprecated)
- GraphTrainableFeatures - is the second plot which users reported using the least.

* similarly it could be good to add as part of the documentation how people can bring their own plots and if there are any missing functionality for the plots we have.

**Comment by anaprietonem** [comment]
Closing this as the refactor is now completed. There is no rendering_backend due to runtime performance differences seen between earthkit-plots and datashader. This feature might be revisited in the future.

**Comment by anaprietonem** [comment]
If in the future any want want to look at the backend rendering option here is the snippet for bencharmking datashader, matplotib and earthkit-plots
```
import numpy as np
import torch
import time
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import datashader as ds
from datashader.mpl_ext import dsshow
import pandas as pd
import earthkit.plots as ekp
import cartopy.crs as ccrs
import cartopy.feature as cfeature

# ── Load graph ──
graph_9km = torch.load(
    '../graph_enc_proc_dec_n320_multiscale_7.pt',
    weights_only=False, map_location="cpu"
)

lats = graph_9km["data"].x[:, 0].cpu().numpy()
lons = graph_9km["data"].x[:, 1].cpu().numpy()
n_nodes = graph_9km["data"].x.shape[0]
lats_deg = np.rad2deg(lats)
lons_deg = np.rad2deg(lons)
print(f"Grid nodes: {n_nodes:,}")

field = (
    50000
    + 3000 * np.sin(lats)
    + 800  * np.cos(2 * lons) * np.cos(lats)
    + 400  * np.sin(3 * lons) * np.cos(lats)
    + 100  * np.random.randn(n_nodes)
)

df = pd.DataFrame({"x": lons_deg, "y": lats_deg, "val": field})
cmap = "RdBu_r"
norm = mcolors.Normalize(vmin=45000, vmax=55000)
n_pixels = 600
N = 5

def save_fig(fig, path):
    """Anemoi-core saving mechanism."""
    fig.canvas.draw()
    image_array = np.array(fig.canvas.renderer.buffer_rgba())
    plt.imsave(path, image_array, dpi=100)
    plt.close(fig)


# ── METHOD 1: dsshow + cartopy PlateCarree ──
dsshow_times = []
for i in range(N):
    t0 = time.perf_counter()
    fig, ax = plt.subplots(figsize=(16, 8),
                           subplot_kw={"projection": ccrs.PlateCarree()})
    ax.set_global()
    dsshow(
        df,
        ds.Point("x", "y"),
        ds.mean("val"),
        cmap=cmap,
        plot_width=n_pixels,
        plot_height=n_pixels,
        norm=norm,
        aspect="auto",
        ax=ax,
    )
    ax.add_feature(cfeature.COASTLINE, linewidth=0.5)
    ax.gridlines(draw_labels=False, linewidth=0.4, linestyle="--", color="gray")
    save_fig(fig, f"/tmp/dsshow_run{i}.png")
    dsshow_times.append(time.perf_counter() - t0)

# ── METHOD 2: earthkit — let it own the figure, same save mechanism ──
ekp_times = []
for i in range(N):
    t0 = time.perf_counter()
    chart = ekp.Map(domain="global")  # PlateCarree by default for "global"
    chart.point_cloud(field, x=lons_deg, y=lats_deg, vmin=45000, vmax=55000, cmap=cmap)
    chart.coastlines(linewidth=0.5)
    chart.gridlines(draw_labels=False, linewidth=0.4, linestyle="--", color="gray")
    # Resize to match dsshow figure
    chart.fig.set_size_inches(16, 8)
    save_fig(chart.fig, f"/tmp/ekp_run{i}.png")
    ekp_times.append(time.perf_counter() - t0)

# ── Results ──
print(f"\n{'Method':<35} {'Mean (s)':>10} {'Min (s)':>10} {'Max (s)':>10}")
print("-" * 67)
print(f"{'dsshow + PlateCarree':<35} {np.mean(dsshow_times):>10.3f} {np.min(dsshow_times):>10.3f} {np.max(dsshow_times):>10.3f}")
print(f"{'earthkit point_cloud':<35} {np.mean(ekp_times):>10.3f} {np.min(ekp_times):>10.3f} {np.max(ekp_times):>10.3f}")
print(f"\ndsshow   — run 0 (JIT): {dsshow_times[0]:.3f}s  |  runs 1-4 mean: {np.mean(dsshow_times[1:]):.3f}s")
print(f"earthkit — run 0:       {ekp_times[0]:.3f}s  |  runs 1-4 mean: {np.mean(ekp_times[1:]):.3f}s")
print(f"\nSpeedup post-warmup: {np.mean(ekp_times[1:])/np.mean(dsshow_times[1:]):.2f}x  (>1 = dsshow faster, <1 = earthkit faster)")

```

<img width="1714" height="472" alt="Image" src="https://github.com/user-attachments/assets/53cfc509-a3e5-42fd-9ef8-8a7400f5a77f" />
